### PR TITLE
feat: add async model average algorithm

### PR DESCRIPTION
**Pull request by NOBLES5E**

### CODE REVIEW

**Comment by NOBLES5E** [comment]
@wangraying fix CI so that we can merge this
We also need a tutorial page on this algorithm

**Comment by wangraying** [comment]
> @wangraying fix CI so that we can merge this
> We also need a tutorial page on this algorithm

OK. I'll add a tutorial page after code is completely ok.

This algorithm will hang on exit in some rare cases. I'm now trying to diagnose the root cause with `nccl` developers.

This pr is now a draft.

**Comment by wangraying** [comment]
update @NOBLES5E 

It seems using `LL128` protocol will cause the problem to hang. I have reproduced the problem through a sample code backend by torch+cupy. This is consistent with the results on Bagua.

The other two protocols `LL` and `Simple` does not hang as far as I know.

I have posted an [issue](https://github.com/NVIDIA/nccl/issues/549) on github, which we could track later on.

**Comment by wangraying** [comment]
> @wangraying fix CI so that we can merge this
> We also need a tutorial page on this algorithm

This CI seems broken.

**Comment by wangraying** [comment]
And do you have any further comment about the usage, since we add a function `barrier` to end async threads.

File: bagua/torch_api/algorithms/async_model_average.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+    def init_operations(
+        self, bagua_module: BaguaModule, bucket: BaguaBucket,
+    ):
+        bucket.clear_ops()
+        bucket.append_asynchronous_model_average_op(
+            peer_selection_mode=self.peer_selection_mode,
+        )
+
+    def abort(self, bagua_module: BaguaModule, grace_period_secs=5):
```
Comment (NOBLES5E) [inline bagua/torch_api/algorithms/async_model_average.py:91]: I changed the function name to `abort`, please check other places

File: bagua/torch_api/algorithms/async_model_average.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            peer_selection_mode=self.peer_selection_mode,
+        )
+
+    def abort(self, bagua_module: BaguaModule, grace_period_seconds=5):
+        """
+        Gracefully stop all workers.
+
+        Args:
+            bagua_module: A PyTorch module initialized by ``with_bagua(...)`` method.
+            grace_period_seconds: Number of seconds a worker will wait before aborting its unfinished communication operations.
```
Comment (NOBLES5E) [inline bagua/torch_api/algorithms/async_model_average.py:97]: this argument is changed to `grace_period_seconds`, please check other places

**Review by NOBLES5E — COMMENTED** [review]
I refined the texts and left some comments. @wangraying Please do a final check

**Pull request merged**